compiler/util/src/org/jetbrains/kotlin/utils/SmartHashMap.kt

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.utils
+
+const val ARRAY_UNTIL_SIZE = 10
+
+/**
+ * [SmartHashMap] is a Map implementation that uses reference identity for keys.
```

Comment:
Also add a note that this class is not thread-safe

Reply:
done